instrumentation/http/src/main/java/brave/http/HttpHandler.java

```diff
@@ -13,68 +13,66 @@
  */
 package brave.http;
 
+import brave.ErrorParser;
 import brave.Span;
+import brave.SpanCustomizer;
 import brave.internal.Nullable;
-import brave.propagation.CurrentTraceContext;
-import brave.propagation.CurrentTraceContext.Scope;
 
 abstract class HttpHandler {
-  final CurrentTraceContext currentTraceContext;
-  final HttpParser parser;
+  final HttpRequestParser requestParser;
+  final HttpResponseParser responseParser;
+  final ErrorParser errorParser;
 
-  HttpHandler(CurrentTraceContext currentTraceContext, HttpParser parser) {
-    this.currentTraceContext = currentTraceContext;
-    this.parser = parser;
+  HttpHandler(
+    HttpRequestParser requestParser,
+    HttpResponseParser responseParser,
+    ErrorParser errorParser
+  ) {
+    this.requestParser = requestParser;
+    this.responseParser = responseParser;
+    this.errorParser = errorParser;
   }
 
-  <Req> Span handleStart(HttpAdapter<Req, ?> adapter, Req request, Span span) {
+  Span handleStart(HttpRequest request, Span span) {
     if (span.isNoop()) return span;
-    Scope ws = currentTraceContext.maybeScope(span.context());
+
+    span.kind(request.spanKind());
     try {
-      parseRequest(adapter, request, span);
+      parseRequest(request, span);
     } finally {
-      ws.close();
-    }
-
-    // all of the above parsing happened before a timestamp on the span
-    long timestamp = adapter.startTimestamp(request);
-    if (timestamp == 0L) {
-      span.start();
-    } else {
-      span.start(timestamp);
+      // all of the above parsing happened before a timestamp on the span
+      long timestamp = request.startTimestamp();
+      if (timestamp == 0L) {
+        span.start();
+      } else {
+        span.start(timestamp);
+      }
     }
     return span;
   }
 
-  /** parses remote IP:port and tags while the span is in scope (for logging for example) */
-  abstract <Req> void parseRequest(HttpAdapter<Req, ?> adapter, Req request, Span span);
+  abstract void parseRequest(HttpRequest request, Span span);
 
+  void handleFinish(@Nullable HttpResponse response, @Nullable Throwable error, Span span) {
+    if (response == null && error == null) {
```

Comment:
@anuraaga this is the check that was moved from `HttpClientHandler#handleReceive`, but I'm also not sure if this check should be moved back to the beginning of `handleReceive`

Reply:
less work is happening in clientHandler.handleReceive, serverHandler.handleSend now... this allows us to avoid copy/paste